xray_core/src/buffer.rs

```diff
@@ -1,20 +1,70 @@
 use std::cmp;
 use std::collections::HashSet;
+use std::fmt::{Debug, Error, Formatter};
 use std::iter;
 use std::ops::{AddAssign, Range};
 use std::sync::Arc;
+use std::sync::mpsc::TrySendError;
+use multiqueue::{MPMCFutReceiver, MPMCFutSender, mpmc_fut_queue};
+use futures::{Async, Poll, Stream};
 use super::tree::{self, Tree};
 
 pub type ReplicaId = usize;
 type LocalTimestamp = usize;
 type LamportTimestamp = usize;
+type Version = LamportTimestamp;
+
+struct MPMCFutQueue<T> {
+    receiver: MPMCFutReceiver<T>,
+    sender: MPMCFutSender<T>
+}
+
+impl<T> MPMCFutQueue<T> {
+    fn new() -> Self {
+        let (sender, receiver) = mpmc_fut_queue(512);
+        MPMCFutQueue { sender, receiver }
+    }
+}
+
+impl<T> Debug for MPMCFutQueue<T> {
+    fn fmt(&self, formatter: &mut Formatter) -> Result<(), Error> {
+        formatter.write_str("MPMCFutQueue").unwrap();
+        Ok(())
+    }
+}
+
+pub struct RegisterStream<T: Stream>(T);
+
+impl<T: Stream> Stream for RegisterStream<T> {
+    type Item = T::Item;
+    type Error = T::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        let mut last_value;
+
+        match self.0.poll()? {
+            Async::Ready(v) => last_value = v,
+            Async::NotReady => return Ok(Async::NotReady)
+        }
+
+        loop {
+            match self.0.poll()? {
+                Async::Ready(v) => last_value = v,
+                Async::NotReady => break
+            }
+        }
+
+        Ok(Async::Ready(last_value))
+    }
+}
 
 #[derive(Debug)]
 pub struct Buffer {
     replica_id: ReplicaId,
     local_clock: LocalTimestamp,
     lamport_clock: LamportTimestamp,
-    fragments: Tree<Fragment>
+    fragments: Tree<Fragment>,
+    change_events_queue: MPMCFutQueue<LamportTimestamp>
```

Comment:
What about two fields here, `changes_sink` and `changes`? `changes_sink` will only be used internally in `broadcast_changes()` when we update the buffer. `changes` can be cloned and returned by a method named `changes()`.